### PR TITLE
start: show warning when using __ACCELERATE and VECLIB_MAXIMUM_THREADS not set

### PR DESCRIPTION
**Pull request by dev-zero**

### CODE REVIEW

File: src/start/cp2k.F [hunk 1]
```diff
@@ -125,6 +126,25 @@ PROGRAM cp2k
       CALL omp_set_num_threads(1)
    ENDIF
 
+#ifdef __ACCELERATE
+   IF (omp_get_max_threads() > 1) THEN
```
Comment (ltalirz) [inline src/start/cp2k.F:130]: Just to understand: why should this not apply when `OMP_NUM_THREADS=1`?

**Comment by dev-zero** [comment]
The veclib threading is not per-se bad, only when we have OMP threading enabled at the same time and then leading to more threads than processors. If OMP threading is disabled it is up to the user and Apple whether or not the veclib threading makes sense.

**Comment by ltalirz** [comment]
It's just that the example I mentioned was exactly with `OMP_NUM_THREADS=1`.
In that case - at least from the reports of the OS - the process is limited to use <=100% of a core, but the threads will still go up to, e.g., 9 (sharing those 100%) and it was less efficient than limiting the veclib threads to 1 as well.

Anyhow, I'll find some time to do a bit more extensive testing at some point

**Comment by dev-zero** [comment]
While that's an interesting case and requires more investigation it isn't really CP2K's problem (I couldn't reproduce it with `QS` benchmark, tried with different combinations of `OMP_NUM_THREADS` and `VECLIB_MAXIMUM_THREADS`, the only case where time/SCF was higher was with both OMP and veclib threads enabled).
If Apple decides to use threads all the time to make computation go Brrrrr.... then that's that. We cannot take into account all the specifics that the foundation library devs come up with.

**Comment by ltalirz** [comment]
> (I couldn't reproduce it with `QS` benchmark, tried with different combinations of `OMP_NUM_THREADS` and `VECLIB_MAXIMUM_THREADS`

I now tried reproducing it myself using a larger example, and indeed the evidence is rather weak (this is an `ENERGY` calculation .

I'm using a large cutoff to have large grids and make things slow but perhaps that just shifts focus to the FFT (?). Which parameters should one tweak to maximize blas/lapack usage?)

[coords.txt](https://github.com/cp2k/cp2k/files/5898479/coords.txt)
[input.txt](https://github.com/cp2k/cp2k/files/5898518/input.txt)

### 4 OMP threads
With a bit of good will, one might make out a difference, at least in the first two steps.
Anyhow, if it is real, it's small.

#### `VECLIB_MAXIMUM_THREADS=1`
```
 Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   12.7     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   15.6     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   15.7     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   15.4     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   15.5     0.01552277      -827.9109704705 -2.49E+0
```
retry
```
  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   12.7     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   15.3     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   15.8     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   15.6     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   15.4     0.01552277      -827.9109704705 -2.49E+0
```
#### `VECLIB_MAXIMUM_THREADS` unlimited
```
 Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   13.4     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   16.5     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   15.3     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   15.7     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   15.6     0.01552277      -827.9109704705 -2.49E+00
```
retry
```
  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   13.6     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   15.7     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   16.3     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   15.4     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   15.6     0.01552277      -827.9109704705 -2.49E+00
```


### 1 OMP thread
I also tried with one thread; here the results are less distinguishable (and perhaps even the other way around).


#### `VECLIB_MAX_THREADS=1`
```
  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   26.0     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   33.7     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   32.7     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   32.8     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   32.8     0.01552277      -827.9109704705 -2.49E+00
```
retry
```
  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   25.5     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   33.7     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   32.6     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   33.3     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   33.4     0.01552277      -827.9109704705 -2.49E+00
```


#### `VECLIB_MAX_THREADS` unlimited
```
  Step     Update method      Time    Convergence         Total energy    Change
  ------------------------------------------------------------------------------
     1 OT DIIS     0.15E+00   25.8     0.05052480      -808.5294683066 -8.09E+02
     2 OT DIIS     0.15E+00   32.9     0.05511683      -817.7789903499 -9.25E+00
     3 OT DIIS     0.15E+00   32.8     0.04309610      -822.7520049686 -4.97E+00
     4 OT DIIS     0.15E+00   32.6     0.03840586      -825.4229805487 -2.67E+00
     5 OT DIIS     0.15E+00   32.8     0.01552277      -827.9109704705 -2.49E+00
```

### Conclusion

Introducing the warning seems like a good idea (and it's fine not to put a warning for `OMP_NUM_THREADS=1`).
I can imagine that things deteriorate when using more and more OMP threads.

**Comment by oschuett** [comment]
I'm wondering if we should also show this warning for other math libraries, e.g. `MKL_NUM_THREADS` or `OPENBLAS_NUM_THREADS`?

Furthermore, we should include a matching remark in #1348.

**Comment by dev-zero** [comment]
> I'm wondering if we should also show this warning for other math libraries, e.g. `MKL_NUM_THREADS` or `OPENBLAS_NUM_THREADS`?

For MKL we already recommend (at least by example and the toolchain) to use the sequential variant. For OpenBLAS we require (again by example and the toolchain) the OMP-variant. The OMP standard should ensure nested threading "just works".

> Furthermore, we should include a matching remark in #1348.

Yes, definitely.

**Pull request merged**